File: MATLAB/correlation/stats_corr_partialCorrs.m
%%
%   COURSE: Master statistics and machine learning: intuition, math, code										
%      URL: udemy.com/course/statsml_x/?couponCode=202304 
% 
%  SECTION: Correlation
%    VIDEO: Partial correlations
% 
%  TEACHER: Mike X Cohen, sincxpress.com
%

%%

% a clear MATLAB workspace is a clear mental workspace
close all; clear; clc

%% the example from the video

% raw correlations
rmg = .7;
rsg = .8;
rms = .9;

% partial correlations
rho_mg_s = (rmg - rsg*rms) / ( sqrt(1-rsg^2)*sqrt(1-rms^2) )
rho_sg_m = (rsg - rmg*rms) / ( sqrt(1-rmg^2)*sqrt(1-rms^2) )

%% now for datasets

N = 76;

% correlated datasets
x1 = linspace(1,10,N)' + randn(N,1);
x2 = x1 + randn(N,1);
x3 = x1 + randn(N,1);


% compute the "raw" correlation matrix
[cormatR,cormatP] = corr([x1 x2 x3]);

% inspect some partial correlations
partialcorr(x2,x3,x1)

%% visualize

figure(1), clf

% raw correlations
subplot(121)
imagesc(cormatR), axis square
set(gca,'clim',[-1 1],'xtick',1:3,'ytick',1:3)
xlabel('Channels'), ylabel('Channels')
title('Raw correlation matrix')

% add text correlation values
for i=1:3
    for j=1:3
        text(i,j,num2str(round(cormatR(i,j),2)),'HorizontalAlignment','center','fontsize',15)
    end
end



% partial correlations
parcormat = partialcorr([x1 x2 x3]);

subplot(122)
imagesc(parcormat), axis square
set(gca,'clim',[-1 1],'xtick',1:3,'ytick',1:3)
xlabel('Channels'), ylabel('Channels')
title('Partial correlation matrix')

% add text correlation values
[I,J] = meshgrid(1:3);
text(I(:),J(:),num2str(round(parcormat(:),2)),'HorizontalAlignment','center','FontSize',15)

%% done.

File: Python/correlation/stats_corr_partialCorrs.py
#!/usr/bin/env python
# coding: utf-8

# # COURSE: Master statistics and machine learning: Intuition, Math, code
# ##### COURSE URL: udemy.com/course/statsml_x/?couponCode=202304 
# ## SECTION: Correlation
# ### VIDEO: Partial correlations
# #### TEACHER: Mike X Cohen, sincxpress.com

# In[ ]:


# import libraries
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scipy.stats as stats


# In[ ]:


# I'm using pingouin for partial correlations.
# You might need to install it, e.g., using the line below.
# This needs to be run only once per install.
# conda install -c conda-forge pingouin
import pingouin as pg


# In[ ]:


## the example from the video

# raw correlations
rmg = .7
rsg = .8
rms = .9

# partial correlations
rho_mg_s = (rmg - rsg*rms) / ( np.sqrt(1-rsg**2)*np.sqrt(1-rms**2) )
rho_sg_m = (rsg - rmg*rms) / ( np.sqrt(1-rmg**2)*np.sqrt(1-rms**2) )

print(rho_mg_s)
print(rho_sg_m)


# In[ ]:


## now for datasets

N = 76

# correlated datasets
x1 = np.linspace(1,10,N) + np.random.randn(N)
x2 = x1 + np.random.randn(N)
x3 = x1 + np.random.randn(N)

# let's convert these data to a pandas frame
df = pd.DataFrame()
df['x1'] = x1
df['x2'] = x2
df['x3'] = x3

# compute the "raw" correlation matrix
cormatR = df.corr()
print(cormatR)

# print out one value
print(' ')
print(cormatR.values[1,0])

# partial correlation
pc = pg.partial_corr(df,x='x3',y='x2',covar='x1')
print(' ')
print(pc)


# In[ ]:


## visualize the matrices

fig,ax = plt.subplots(1,2,figsize=(6,3))

# raw correlations
ax[0].imshow(cormatR.values,vmin=-1,vmax=1)
ax[0].set_xticks(range(3))
ax[0].set_yticks(range(3))

# add text 
for i in range(3):
    for j in range(3):
        ax[0].text(i,j,np.round(cormatR.values[i,j],2), horizontalalignment='center')

        
        
# partial correlations
partialCorMat = df.pcorr()
ax[1].imshow(partialCorMat.values,vmin=-1,vmax=1)
ax[1].set_xticks(range(3))
ax[1].set_yticks(range(3))

for i in range(3):
    for j in range(3):
        ax[1].text(i,j,np.round(partialCorMat.values[i,j],2), horizontalalignment='center')


plt.show()

